pcdsdevices/attenuator.py

```diff
@@ -1071,6 +1072,114 @@ class AT2L0(FltMvInterface, PVPositionerPC, LightpathInOutMixin):
     # Summary for lightpath view
     num_in = Cpt(InternalSignal, kind='hinted')
     num_out = Cpt(InternalSignal, kind='hinted')
+    
+    """check for errors and return a string indicating any errors verbally"""
+    def _check_errors(signals: SignalToValue) ->  str:
+        errors = []
+        #sort out .motor from .motor.plc signals
+        for sig, value in signals.items():
+            if ("motor" in sig.name) ^ ("plc" in sig.name):
+                value=sig.metadata.get('severity').value
+
+            if value not in (0,""):
+                errors.append(f"{sig.name}: {value}")
+
+        if errors:
+            return "\n".join(["Error summary:"] + errors)
+
+        return "No errors"
+
+    """summarize the errors occurring at any time on any blade via a string"""
+    error_summary = Cpt(
+        MultiDerivedSignalRO,
+        calculate=_check_errors,
+        attrs=sum(
```

Comment:
A couple things here:
1. I think you make this attribute list twice. You might consider defining it once and reusing it.
2. Alternatively, despite me being the originator of this ugly chunk of code in the other PR, let me suggest a more readable way of generating this list - a [generator](https://wiki.python.org/moin/Generators):

```python
def _get_blade_error_attrs():
    """Get the blade attribute names used for checking errors."""
    for index in range(1,20)
        yield f"blade_{index:02d}.state.error"
        yield f"blade_{index:02d}.state.error_id"
        yield f"blade_{index:02d}.state.error_message"
        yield f"blade_{index:02d}.motor.plc.err_code"
        yield f"blade_{index:02d}.motor.user_readback"
```

Then this component definition becomes

```python
    error_summary = Cpt(
        MultiDerivedSignalRO,
        calculate=_check_errors,
        attrs=list(_get_blade_error_attrs()),
    )
```

I think we could probably agree the above is a bit more readable...

Reply:
oh interesting, is this generator pattern pythons take on recursion?

Reply:
looking a little closer, they aren't the same, but seemingly similar concepts of repeatedly calling a function to enact iteration.

Reply:
It's certainly a python-ism.  It's not really recursion, but it's a way of dealing with iteration efficiently.   In short, a generator gives you an Iterable object that "yields" its items one at a time.   (rather than having to gather them all at once like a list would).

It's also something that shows up in Bluesky quite often, so I'm sure you'll be come more familiar with it soon. 😄